### PR TITLE
Add generic type param for addReplaceFile.

### PR DESCRIPTION
**Pull request by TwitchBronBron**
**Notable changes:**
 - define new type called `BscFile` which is the union of `BrsFile` and `XmlFile` to simplify types across the project.
 - Add generic type parameter for the "add file(s)" functions on `Program` to simplify code and tests.

### CODE REVIEW

File: src/Program.ts [hunk 1]
```diff
@@ -235,14 +235,14 @@ export class Program {
      * contents from the file system.
      * @param filePaths
      */
-    public async addOrReplaceFiles(fileObjects: Array<FileObj>) {
+    public async addOrReplaceFiles<T extends BscFile[]>(fileObjects: Array<FileObj>) {
```
Comment (TwitchBronBron) [inline src/Program.ts:238]: Add generic type param

File: src/Program.ts [hunk 2]
```diff
@@ -284,15 +284,15 @@ export class Program {
      * @param relativePath the file path relative to the root dir
      * @param fileContents the file contents. If not provided, the file will be loaded from disk
      */
-    public async addOrReplaceFile(relativePath: string, fileContents?: string): Promise<XmlFile | BrsFile>;
+    public async addOrReplaceFile<T extends BscFile>(relativePath: string, fileContents?: string): Promise<T>;
```
Comment (TwitchBronBron) [inline src/Program.ts:287]: Add generic type param

File: src/Program.ts [hunk 3]
```diff
@@ -424,7 +424,7 @@ export class Program {
      * Get a list of files for the given pkgPath array.
      * Missing files are just ignored.
      */
-    public getFilesByPkgPaths(pkgPaths: string[]) {
+    public getFilesByPkgPaths<T extends BscFile[]>(pkgPaths: string[]) {
```
Comment (TwitchBronBron) [inline src/Program.ts:427]: Add generic type param

File: src/Program.ts [hunk 4]
```diff
@@ -434,15 +434,15 @@ export class Program {
...
+        return result as T;
     }
 
     /**
      * Get a file with the specified pkg path.
      * If not found, return undefined
      */
-    public getFileByPkgPath(pkgPath: string) {
-        return this.getFilesByPkgPaths([pkgPath])[0];
+    public getFileByPkgPath<T extends BscFile>(pkgPath: string) {
```
Comment (TwitchBronBron) [inline src/Program.ts:444]: Add generic type param

File: src/Program.ts [hunk 5]
```diff
@@ -586,9 +586,9 @@ export class Program {
      * Get the file at the given path
      * @param pathAbsolute
      */
-    private getFile(pathAbsolute: string) {
+    private getFile<T extends BscFile>(pathAbsolute: string) {
```
Comment (TwitchBronBron) [inline src/Program.ts:589]: Add generic type param

Comment (elsassph) [reply src/Program.ts:589]: Does it need to be generic?

Comment (elsassph) [reply src/Program.ts:589]: Ah got it from usage.

**Pull request merged**